### PR TITLE
Use Index/Address/Offset instead of uint32_t

### PR DESCRIPTION
**Pull request by binji**
An `Index` is an index into one of the WebAssembly index spaces. It also
is used for counts for these spaces, as well as parameter counts and
result counts.

An `Address` is an index into linear memory, or the size of a data
region in linear memory.

An `Offset` is an offset into the host's file or memory buffer.

This fixes issue #322.

### CODE REVIEW

**Comment by sbc100** [comment]
Can you explain why you used Index for things like param counts?   This doesn't really seem like a index space in the same why... but either way this change seem like an improvement.

File: src/wat-writer.cc [hunk 1]
```diff
@@ -122,14 +122,14 @@ class WatWriter {
   Result result_ = Result::Ok;
   int indent_ = 0;
   NextChar next_char_ = NextChar::None;
-  int depth_ = 0;
+  Index depth_ = 0;
```
Comment (sbc100) [inline src/wat-writer.cc]: Why is depth_ an index?

Comment (binji) [inline src/wat-writer.cc]: Depth is an index into the label index space, see: http://webassembly.github.io/spec/syntax/modules.html#indices

**Comment by binji** [comment]
> Can you explain why you used Index for things like param counts?

Well, the spec uses `localidx` for accessing via `get_local` etc., so it seemed reasonable. I was considering using a different typedef for counts vs. indexes (for example using `Address` for linear memory sizes seems a bit weird), but I don't know that it would actually make things clearer.

**Pull request merged**